### PR TITLE
feat(connector): [NMI] Currency Unit Conversion

### PR DESCRIPTION
**Pull request by SagarDevAchar**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
- Addressing Issue: #2236 
- Modified two files in `hyperswitch/crates/router/src/connector/`
  - `nmi.rs`
    - Implement `get_currency_unit` function
    - Modify `ConnectorIntegration` implementations for `Nmi`
  - `nmi/transformers.rs`
    - Implement `NmiRouterData<T>` structure and functionality
    - Modify implementation for `NmiPaymentsRequest`

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

We need to create a payment using NMI and test the currency unit from connector response and connector dashboard.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/nmi/transformers.rs [hunk 1]
```diff
@@ -40,6 +40,37 @@ impl TryFrom<&ConnectorAuthType> for NmiAuthType {
...
+    fn try_from(
+        (_currency_unit, _currency, amount, router_data): (
+            &types::api::CurrencyUnit,
+            types::storage::enums::Currency,
+            i64,
+            T,
+        ),
+    ) -> Result<Self, Self::Error> {
+        Ok(Self {
+            amount,
```
Comment (srujanchikke) [inline crates/router/src/connector/nmi/transformers.rs:68]:
```suggestion
            amount: utils::to_currency_base_unit_asf64(_amount, _currency)?,)
```
amount needs to f64 as connector accepts amount it in base unit

File: crates/router/src/connector/nmi/transformers.rs [hunk 2]
```diff
@@ -76,23 +107,28 @@ pub struct ApplePayData {
...
         };
-        let auth_type: NmiAuthType = (&item.connector_auth_type).try_into()?;
-        let amount =
-            utils::to_currency_base_unit_asf64(item.request.amount, item.request.currency)?;
-        let payment_method = PaymentMethod::try_from(&item.request.payment_method_data)?;
+        let auth_type: NmiAuthType = (&item.router_data.connector_auth_type).try_into()?;
+        let amount = utils::to_currency_base_unit_asf64(
+            item.router_data.request.amount,
+            item.router_data.request.currency,
+        )?;
```
Comment (srujanchikke) [inline crates/router/src/connector/nmi/transformers.rs:123]:
```suggestion
        let amount =item.amount;
```

File: crates/router/src/connector/nmi/transformers.rs [hunk 3]
```diff
@@ -574,17 +610,17 @@ pub struct NmiRefundRequest {
...
-    fn try_from(item: &types::RefundsRouterData<F>) -> Result<Self, Self::Error> {
-        let auth_type: NmiAuthType = (&item.connector_auth_type).try_into()?;
+    fn try_from(item: &NmiRouterData<&types::RefundsRouterData<F>>) -> Result<Self, Self::Error> {
+        let auth_type: NmiAuthType = (&item.router_data.connector_auth_type).try_into()?;
         Ok(Self {
             transaction_type: TransactionType::Refund,
             security_key: auth_type.api_key,
-            transactionid: item.request.connector_transaction_id.clone(),
+            transactionid: item.router_data.request.connector_transaction_id.clone(),
             amount: utils::to_currency_base_unit_asf64(
```
Comment (srujanchikke) [inline crates/router/src/connector/nmi/transformers.rs:621]:
```suggestion
            amount: item.amount ,
```

**Review by srujanchikke — CHANGES_REQUESTED** [review]
Capture flow is also using amount field , Could you also make similar changes in capture try from ?

File: crates/router/src/connector/nmi/transformers.rs [hunk 4]
```diff
@@ -243,17 +276,19 @@ pub struct NmiCaptureRequest {
...
+    fn try_from(
+        item: &NmiRouterData<&types::PaymentsCaptureRouterData>,
+    ) -> Result<Self, Self::Error> {
+        let auth = NmiAuthType::try_from(&item.router_data.connector_auth_type)?;
         Ok(Self {
             transaction_type: TransactionType::Capture,
             security_key: auth.api_key,
-            transactionid: item.request.connector_transaction_id.clone(),
+            transactionid: item.router_data.request.connector_transaction_id.clone(),
             amount: Some(utils::to_currency_base_unit_asf64(
```
Comment (AkshayaFoiger) [inline crates/router/src/connector/nmi/transformers.rs:289]: The variable `amount` should be assigned the value of `item.amount.`

**Review by AkshayaFoiger — CHANGES_REQUESTED** [review]

**Comment by swangi-kumari** [comment]
Hey @SagarDevAchar ,
Thanks for your interest in contributing to hyperswitch.
Let us know if you need any assistance from our end.
Also, even if hacktoberfest is over, we should celebrate open source everyday and we are open for more contributions from you.
We would still be rewarding folks with goodies even if the PR gets merged post hacktoberfest.
May the Source be with you!

File: crates/router/src/connector/nmi/transformers.rs (hunk 4 above)
Comment (SagarDevAchar) [reply crates/router/src/connector/nmi/transformers.rs]: Resolved! Thanks @AkshayaFoiger!

File: crates/router/src/connector/nmi/transformers.rs (hunk 3 above)
Comment (SagarDevAchar) [reply crates/router/src/connector/nmi/transformers.rs]: Resolved! Thanks @srujanchikke!

**Comment by AkshayaFoiger** [comment]
Looks Good!

**Review by AkshayaFoiger — APPROVED** [review]

**Review by srujanchikke — APPROVED** [review]
LGTM!

**Review by SanchithHegde — APPROVED** [review]
Looks good to me!

Thanks for the PR, @SagarDevAchar!

**Comment by deepanshu-iiitu** [comment]
Hey @SagarDevAchar   ,
Thanks a bunch for all your contributions! We've got a little something for you to show our appreciation.
Just take a moment to fill out this [form](https://forms.gle/yqbtjFuP6oxLaBKN8), and get ready for some awesome swag coming your way.
Thanks!

**Pull request merged**